File: py-xbbg/src/xbbg/ext/fixed_income.py
"""Fixed income extension functions.

Convenience wrappers for fixed income and bond analysis queries.
Returns DataFrame in the configured backend format.
Uses high-performance Rust utilities from xbbg._core.

Sync functions (wrap async with asyncio.run):
    - yas(): Yield & Spread Analysis
    - preferreds(): Find preferred stocks for a company
    - corporate_bonds(): Find corporate bonds for a company
    - bqr(): Bloomberg Quote Request (dealer quotes)

Async functions (primary implementation):
    - ayas(): Async yield & spread analysis
    - apreferreds(): Async find preferred stocks
    - acorporate_bonds(): Async find corporate bonds
    - abqr(): Async Bloomberg Quote Request
"""

from __future__ import annotations

from datetime import date
from enum import IntEnum
from typing import TYPE_CHECKING

# Import Rust ext utilities for max performance
from xbbg._core import (
    ext_build_corporate_bonds_query,
    ext_build_preferreds_query,
    ext_build_yas_overrides,
    ext_normalize_tickers,
)
from xbbg.ext._utils import _fmt_date, _syncify

if TYPE_CHECKING:
    from narwhals.typing import IntoDataFrame


class YieldType(IntEnum):
    """Bloomberg YAS yield type flags for YAS_YLD_FLAG override.

    These values control which yield calculation method Bloomberg uses
    in Yield & Spread Analysis (YAS) calculations.

    Standard Bloomberg YAS_YLD_FLAG values:
        1 = Yield to Maturity (YTM)
        2 = Yield to Call (YTC)
        3 = Yield to Refunding (YTR)
        4 = Yield to Next Put (YTP)
        5 = Yield to Worst (YTW)
        6 = Yield to Worst Refunding (YTWR)
        7 = Euro Yield to Worst (EYTW)
        8 = Euro Yield to Worst Refunding (EYTWR)
        9 = Yield to Average Life (YTAL)
    """

    YTM = 1
    YTC = 2
    YTR = 3
    YTP = 4
    YTW = 5
    YTWR = 6
    EYTW = 7
    EYTWR = 8
    YTAL = 9


def _normalize_tickers(tickers: str | list[str]) -> list[str]:
    """Normalize tickers to a list using Rust."""
    # ext_normalize_tickers expects a list, so wrap single string
    if isinstance(tickers, str):
        return ext_normalize_tickers([tickers])
    return ext_normalize_tickers(tickers)


# =============================================================================
# Async implementations (primary)
# =============================================================================


async def ayas(
    tickers: str | list[str],
    flds: str | list[str] = "YAS_BOND_YLD",
    *,
    settle_dt: str | date | None = None,
    yield_type: YieldType | int | None = None,
    spread: float | None = None,
    yield_: float | None = None,
    price: float | None = None,
    benchmark: str | None = None,
    **kwargs,
) -> IntoDataFrame:
    """Async yield and spread analysis for fixed income securities.

    Convenience wrapper around abdp() for Bloomberg's YAS (Yield & Spread Analysis).
    Maps named parameters to Bloomberg YAS override fields.

    Args:
        tickers: Single ticker or list of bond tickers.
        flds: Field(s) to retrieve. Common YAS fields:
            - YAS_BOND_YLD: Calculated yield
            - YAS_YLD_SPREAD: Spread to benchmark
            - YAS_BOND_PX: Calculated price
            - YAS_ASSET_SWP_SPD: Asset swap spread
            - YAS_MOD_DUR: Modified duration
            - YAS_ISPREAD: I-spread
            - YAS_ZSPREAD: Z-spread
            - YAS_OAS: Option-adjusted spread
        settle_dt: Settlement date for the calculation. Default: spot settlement.
        yield_type: Type of yield calculation. Use YieldType enum or int:
            - YieldType.YTM (1): Yield to Maturity
            - YieldType.YTC (2): Yield to Call
            - YieldType.YTR (3): Yield to Refunding
            - YieldType.YTP (4): Yield to Next Put
            - YieldType.YTW (5): Yield to Worst (next put, call, or maturity)
            - YieldType.YTWR (6): Yield to Worst Refunding
            - YieldType.EYTW (7): Euro Yield to Worst
            - YieldType.EYTWR (8): Euro Yield to Worst Refunding
            - YieldType.YTAL (9): Yield to Average Life
        spread: Input spread value (for reverse calculation from spread to price/yield).
        yield_: Input yield value (for reverse calculation from yield to price).
        price: Input price value (for reverse calculation from price to yield/spread).
        benchmark: Benchmark security for spread calculation (e.g., "T 4.5 05/15/38 Govt").
        **kwargs: Additional arguments passed to abdp().

    Returns:
        DataFrame with yield analysis data (type depends on configured backend).

    Example::

        import asyncio
        from xbbg.ext.fixed_income import ayas, YieldType


        async def main():
            # Get yield to maturity for a bond
            df = await ayas("US912810TM69 Govt", "YAS_BOND_YLD")

            # Get multiple YAS fields
            df = await ayas(
                "US912810TM69 Govt",
                ["YAS_BOND_YLD", "YAS_MOD_DUR", "YAS_ZSPREAD"],
            )

            # Calculate price from yield
            df = await ayas(
                "US912810TM69 Govt",
                "YAS_BOND_PX",
                yield_=4.5,
                yield_type=YieldType.YTM,
            )


        asyncio.run(main())
    """
    from xbbg import abdp

    tickers_list = _normalize_tickers(tickers)

    # Normalize fields
    if isinstance(flds, str):
        fields_list = [flds]
    else:
        fields_list = list(flds)

    # Build overrides using Rust (high performance)
    formatted_dt = _fmt_date(settle_dt) if settle_dt is not None else None
    yt_flag = int(yield_type) if yield_type is not None else None

    yas_pairs = ext_build_yas_overrides(
        settle_dt=formatted_dt,
        yield_type=yt_flag,
        spread=spread,
        yield_val=yield_,
        price=price,
        benchmark=benchmark,
    )
    overrides: dict[str, str] = dict(yas_pairs)

    # Merge with any additional overrides from kwargs
    if "overrides" in kwargs:
        existing = kwargs.pop("overrides")
        if isinstance(existing, dict):
            overrides.update(existing)
        elif isinstance(existing, list):
            for k, v in existing:
                overrides[k] = str(v)

    # Call abdp with the overrides
    return await abdp(tickers=tickers_list, flds=fields_list, overrides=overrides, **kwargs)


async def apreferreds(
    equity_ticker: str,
    *,
    fields: list[str] | None = None,
    **kwargs,
) -> IntoDataFrame:
    """Async find preferred stocks for a company using BQL.

    Uses Bloomberg's debt filter to find preferred stock issues
    associated with a given equity ticker.

    Args:
        equity_ticker: Company equity ticker (e.g., "BAC US Equity" or "BAC").
            If no suffix is provided, " US Equity" will be appended.
        fields: Optional list of additional fields to retrieve.
            Default fields are: id, name.
        **kwargs: Additional options passed to the underlying BQL query.

    Returns:
        DataFrame with preferred stock information (type depends on configured backend).
        Columns include the security ID, name, and any additional requested fields.

    Example::

        import asyncio
        from xbbg.ext.fixed_income import apreferreds


        async def main():
            # Get preferred stocks for Bank of America
            df = await apreferreds("BAC US Equity")

            # Get preferreds with additional fields
            df = await apreferreds("BAC", fields=["px_last", "dvd_yld"])


        asyncio.run(main())
    """
    from xbbg import abql

    # Build BQL query using Rust (handles ticker normalization, field dedup)
    extra = list(fields) if fields else []
    bql_query = ext_build_preferreds_query(equity_ticker, extra)

    return await abql(bql_query, **kwargs)


async def acorporate_bonds(
    ticker: str,
    *,
    ccy: str | None = "USD",
    fields: list[str] | None = None,
    active_only: bool = True,
    **kwargs,
) -> IntoDataFrame:
    """Async find corporate bonds for a company using BQL.

    Uses Bloomberg's debt() universe to find corporate bond issues
    for a given company via its equity ticker. Works across all markets.

    Args:
        ticker: Company equity ticker (e.g., "AAPL", "9984 JT Equity").
            If no suffix is provided, " US Equity" is appended.
        ccy: Currency filter (default: "USD"). Set to None for all currencies.
        fields: Optional list of additional fields to retrieve.
            Default field is: id.
        active_only: If True (default), only return active bonds.
        **kwargs: Additional options passed to the underlying BQL query.

    Returns:
        DataFrame with corporate bond information (type depends on configured backend).
        Columns include the security ID and any additional requested fields.

    Example::

        import asyncio
        from xbbg.ext.fixed_income import acorporate_bonds


        async def main():
            # Get active USD corporate bonds for Apple
            df = await acorporate_bonds("AAPL")

            # Get all currency bonds with additional fields
            df = await acorporate_bonds("MSFT", ccy=None, fields=["name", "cpn", "maturity"])


        asyncio.run(main())
    """
    from xbbg import abql

    # Build BQL query using Rust (handles field dedup, filter construction)
    extra = list(fields) if fields else []
    bql_query = ext_build_corporate_bonds_query(ticker, ccy, extra, active_only)

    return await abql(bql_query, **kwargs)


async def abqr(
    ticker: str,
    *,
    start_datetime: str | None = None,
    end_datetime: str | None = None,
    event_types: list[str] | None = None,
    include_broker_codes: bool = True,
    **kwargs,
) -> IntoDataFrame:
    """Async Bloomberg Quote Request (dealer quotes).

    Retrieves intraday tick data with broker/dealer codes for a security.
    This is useful for analyzing dealer activity and market making.

    Note: Broker code data availability depends on your Bloomberg entitlements
    and the security type. Not all securities have broker attribution.

    Args:
        ticker: Security ticker (e.g., "US912810TM69 Govt").
        start_datetime: Start datetime (ISO format or "YYYY-MM-DD HH:MM").
            Default: 1 hour ago.
        end_datetime: End datetime (ISO format or "YYYY-MM-DD HH:MM").
            Default: now.
        event_types: List of event types to retrieve (default: ["BID", "ASK"]).
            Options: "TRADE", "BID", "ASK", "BID_BEST", "ASK_BEST", etc.
        include_broker_codes: Whether to include broker/dealer codes (default: True).
        **kwargs: Additional options passed to abdtick.

    Returns:
        DataFrame with quote data including broker codes (if available).
        Columns typically include: time, type, value, size, and broker codes.

    Example::

        import asyncio
        from xbbg.ext.fixed_income import abqr


        async def main():
            # Get dealer quotes for a government bond
            df = await abqr("US912810TM69 Govt")

            # Get quotes for specific time range
            df = await abqr(
                "US912810TM69 Govt",
                start_datetime="2024-01-15 09:00",
                end_datetime="2024-01-15 10:00",
            )


        asyncio.run(main())
    """
    from xbbg import abdtick
    from xbbg._core import ext_default_bqr_datetimes

    # Compute default datetime range using Rust (handles normalization + defaults)
    start_datetime, end_datetime = ext_default_bqr_datetimes(start_datetime, end_datetime)

    # Default event types
    if event_types is None:
        event_types = ["BID", "ASK"]

    # Add broker code request to kwargs if desired
    if include_broker_codes:
        kwargs["includeBrokerCodes"] = True

    # Use abdtick to get the data
    return await abdtick(
        ticker=ticker,
        start_datetime=start_datetime,
        end_datetime=end_datetime,
        **kwargs,
    )


yas = _syncify(ayas)
preferreds = _syncify(apreferreds)
corporate_bonds = _syncify(acorporate_bonds)
bqr = _syncify(abqr)

File: crates/xbbg-recipes/src/fixed_income.rs
//! Fixed income recipe functions.
//!
//! High-level recipes for Bloomberg fixed income data queries.

use arrow::array::RecordBatch;
use xbbg_async::engine::{Engine, RequestParams};
use xbbg_async::services::{Operation, Service};
use xbbg_ext::transforms::fixed_income::{build_yas_overrides, YieldType};

use crate::error::Result;

/// YAS (Yield & Spread Analysis) recipe.
///
/// Retrieves Bloomberg YAS data with optional yield type and pricing parameters.
///
/// # Arguments
///
/// * `engine` - Bloomberg engine reference
/// * `tickers` - Securities to query
/// * `fields` - Fields to retrieve
/// * `settle_dt` - Settlement date (YYYYMMDD format)
/// * `yield_type` - Yield calculation type (YTM, YTC, etc.)
/// * `spread` - Yield spread override
/// * `yield_val` - Yield value override
/// * `price` - Price override
/// * `benchmark` - Benchmark security for spread calculation
///
/// # Returns
///
/// Arrow RecordBatch with YAS data in canonical long format
///
/// # Example
///
/// ```ignore
/// let batch = recipe_yas(
///     &engine,
///     vec!["US912810SV17 Govt".to_string()],
///     vec!["YAS_BOND_YLD".to_string(), "YAS_YLD_SPREAD".to_string()],
///     Some("20240115".to_string()),
///     Some(YieldType::YTM),
///     None,
///     None,
///     Some(99.5),
///     None,
/// ).await?;
/// ```
#[allow(clippy::too_many_arguments)]
pub async fn recipe_yas(
    engine: &Engine,
    tickers: Vec<String>,
    fields: Vec<String>,
    settle_dt: Option<String>,
    yield_type: Option<YieldType>,
    spread: Option<f64>,
    yield_val: Option<f64>,
    price: Option<f64>,
    benchmark: Option<String>,
) -> Result<RecordBatch> {
    // Build YAS overrides using xbbg-ext helper
    let overrides = build_yas_overrides(
        settle_dt.as_deref(),
        yield_type,
        spread,
        yield_val,
        price,
        benchmark.as_deref(),
    );

    // Build request parameters using canonical enums
    let params = RequestParams {
        service: Service::RefData.to_string(),
        operation: Operation::ReferenceData.to_string(),
        securities: Some(tickers),
        fields: Some(fields),
        overrides: Some(overrides),
        ..Default::default()
    };

    // Call engine directly (no recursion)
    let batch = engine.request(params).await?;
    Ok(batch)
}

/// Find preferred stocks for a company via BQL.
///
/// Uses Bloomberg's debt filter to find preferred stock issues
/// associated with a given equity ticker.
///
/// # Arguments
///
/// * `engine` - Bloomberg engine reference
/// * `equity_ticker` - Company equity ticker (e.g., "BAC US Equity")
/// * `fields` - Fields to retrieve (default: id, name)
///
/// # Returns
///
/// Arrow RecordBatch with preferred stock data
pub async fn recipe_preferreds(
    engine: &Engine,
    equity_ticker: String,
    fields: Option<Vec<String>>,
) -> Result<RecordBatch> {
    // Build field list with defaults
    let all_fields = match fields {
        Some(mut flds) => {
            let mut defaults = vec!["id".to_string(), "name".to_string()];
            defaults.append(&mut flds);
            defaults
        }
        None => vec!["id".to_string(), "name".to_string()],
    };

    let fields_str = all_fields.join(", ");

    // Build BQL query using debt filter with Preferreds asset class
    let bql_query = format!(
        "get({fields_str}) for(filter(debt(['{equity_ticker}'], CONSOLIDATEDUPLICATES='N'), \
         SRCH_ASSET_CLASS=='Preferreds'))"
    );

    let params = RequestParams {
        service: Service::BqlSvc.to_string(),
        operation: Operation::BqlSendQuery.to_string(),
        elements: Some(vec![("expression".to_string(), bql_query)]),
        ..Default::default()
    };

    engine.request(params).await.map_err(Into::into)
}

/// Find corporate bonds for a company via BQL.
///
/// Uses Bloomberg's bondsuniv filter to find active corporate bond issues
/// for a given company ticker, optionally filtered by currency.
///
/// # Arguments
///
/// * `engine` - Bloomberg engine reference
/// * `ticker` - Company ticker prefix (e.g., "AAPL")
/// * `ccy` - Currency filter (e.g., "USD"). None for all currencies.
/// * `fields` - Fields to retrieve (default: id)
/// * `active_only` - If true, only return active bonds
///
/// # Returns
///
/// Arrow RecordBatch with corporate bond data
pub async fn recipe_corporate_bonds(
    engine: &Engine,
    ticker: String,
    ccy: Option<String>,
    fields: Option<Vec<String>>,
    active_only: bool,
) -> Result<RecordBatch> {
    // Build field list with defaults
    let all_fields = match fields {
        Some(mut flds) => {
            let mut defaults = vec!["id".to_string()];
            defaults.append(&mut flds);
            defaults
        }
        None => vec!["id".to_string()],
    };

    let fields_str = all_fields.join(", ");

    // Build filter conditions
    let mut conditions = vec![
        "SRCH_ASSET_CLASS=='Corporates'".to_string(),
        format!("TICKER=='{ticker}'"),
    ];
    if let Some(c) = ccy {
        conditions.push(format!("CRNCY=='{c}'"));
    }
    let filter_str = conditions.join(" AND ");

    let universe = if active_only { "active" } else { "all" };
    let bql_query = format!(
        "get({fields_str}) for(filter(bondsuniv('{universe}', CONSOLIDATEDUPLICATES='N'), {filter_str}))"
    );

    let params = RequestParams {
        service: Service::BqlSvc.to_string(),
        operation: Operation::BqlSendQuery.to_string(),
        elements: Some(vec![("expression".to_string(), bql_query)]),
        ..Default::default()
    };

    engine.request(params).await.map_err(Into::into)
}

/// Bloomberg Quote Request — dealer quotes via IntradayTick.
///
/// Retrieves intraday tick data with broker/dealer codes for a security.
/// Useful for analyzing dealer activity and market making.
///
/// # Arguments
///
/// * `engine` - Bloomberg engine reference
/// * `ticker` - Security ticker (e.g., "US912810TM69 Govt")
/// * `start_datetime` - Start datetime (ISO format)
/// * `end_datetime` - End datetime (ISO format)
/// * `event_types` - Event types to retrieve (default: BID, ASK)
/// * `include_broker_codes` - Include broker/dealer codes (default: true)
///
/// # Returns
///
/// Arrow RecordBatch with quote data including broker codes
pub async fn recipe_bqr(
    engine: &Engine,
    ticker: String,
    start_datetime: String,
    end_datetime: String,
    event_types: Option<Vec<String>>,
    include_broker_codes: bool,
) -> Result<RecordBatch> {
    let evts = event_types.unwrap_or_else(|| vec!["BID".to_string(), "ASK".to_string()]);

    let mut options = vec![];
    if include_broker_codes {
        options.push(("includeBrokerCodes".to_string(), "true".to_string()));
    }

    let params = RequestParams {
        service: Service::RefData.to_string(),
        operation: Operation::IntradayTick.to_string(),
        security: Some(ticker),
        start_datetime: Some(start_datetime),
        end_datetime: Some(end_datetime),
        event_types: Some(evts),
        options: if options.is_empty() {
            None
        } else {
            Some(options)
        },
        ..Default::default()
    };

    engine.request(params).await.map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_recipe_yas_builds_correct_params() {
        // This test verifies parameter building logic
        // Actual engine calls require Bloomberg connection (integration test)

        let overrides = build_yas_overrides(
            Some("20240115"),
            Some(YieldType::YTM),
            None,
            None,
            Some(99.5),
            None,
        );

        // Verify YAS overrides are built correctly
        assert!(overrides
            .iter()
            .any(|(k, v)| k == "YAS_SETTLE_DT" && v == "20240115"));
        assert!(overrides
            .iter()
            .any(|(k, v)| k == "YAS_YLD_FLAG" && v == "1"));
        assert!(overrides
            .iter()
            .any(|(k, v)| k == "YAS_BOND_PX" && v == "99.5"));
    }

    #[test]
    fn test_recipe_preferreds_default_fields() {
        // Verify that default fields are id and name
        let fields: Option<Vec<String>> = None;
        let all_fields = match fields {
            Some(mut flds) => {
                let mut defaults = vec!["id".to_string(), "name".to_string()];
                defaults.append(&mut flds);
                defaults
            }
            None => vec!["id".to_string(), "name".to_string()],
        };
        assert_eq!(all_fields, vec!["id", "name"]);
    }

    #[test]
    fn test_recipe_preferreds_custom_fields() {
        let fields = Some(vec!["px_last".to_string(), "dvd_yld".to_string()]);
        let all_fields = match fields {
            Some(mut flds) => {
                let mut defaults = vec!["id".to_string(), "name".to_string()];
                defaults.append(&mut flds);
                defaults
            }
            None => vec!["id".to_string(), "name".to_string()],
        };
        assert_eq!(all_fields, vec!["id", "name", "px_last", "dvd_yld"]);
    }

    #[test]
    fn test_recipe_corporate_bonds_filter_building() {
        // Test that filter conditions are built correctly
        let ticker = "AAPL".to_string();
        let ccy = Some("USD".to_string());
        let active_only = true;

        let mut conditions = vec![
            "SRCH_ASSET_CLASS=='Corporates'".to_string(),
            format!("TICKER=='{ticker}'"),
        ];
        if let Some(c) = ccy {
            conditions.push(format!("CRNCY=='{c}'"));
        }
        let filter_str = conditions.join(" AND ");
        let universe = if active_only { "active" } else { "all" };

        assert_eq!(
            filter_str,
            "SRCH_ASSET_CLASS=='Corporates' AND TICKER=='AAPL' AND CRNCY=='USD'"
        );
        assert_eq!(universe, "active");
    }

    #[test]
    fn test_recipe_corporate_bonds_no_ccy() {
        let ticker = "MSFT".to_string();
        let ccy: Option<String> = None;

        let mut conditions = vec![
            "SRCH_ASSET_CLASS=='Corporates'".to_string(),
            format!("TICKER=='{ticker}'"),
        ];
        if let Some(c) = ccy {
            conditions.push(format!("CRNCY=='{c}'"));
        }
        let filter_str = conditions.join(" AND ");

        assert_eq!(
            filter_str,
            "SRCH_ASSET_CLASS=='Corporates' AND TICKER=='MSFT'"
        );
    }

    #[test]
    fn test_recipe_bqr_default_event_types() {
        let evts = resolve_event_types(None);
        assert_eq!(evts, vec!["BID", "ASK"]);
    }

    #[test]
    fn test_recipe_bqr_custom_event_types() {
        let evts = resolve_event_types(Some(vec!["TRADE".to_string()]));
        assert_eq!(evts, vec!["TRADE"]);
    }

    fn resolve_event_types(event_types: Option<Vec<String>>) -> Vec<String> {
        event_types.unwrap_or_else(|| vec!["BID".to_string(), "ASK".to_string()])
    }
}
